_posts/2020-12-25-manual-gates-git-flow-in-a-wig.md

```diff
@@ -0,0 +1,145 @@
+
+# A Flexible Model For Software Delivery Pipelines
+
+Assume we are deploying software-as-a-service. This could be an internal or customer facing service.
+
+1. Customers only consume the service from code that has been merged to the master branch.
+
+2. Features that are ready for consumption are merged to the master branch.
+
+3. We have one pipeline that is 'subscribed' to the master branch.
+
+4. All environments in the pipeline exist for the sole purpose of ensuring that changes do not degrade the customer-facing instance.
+
+5. Developers may create a simplified version of the master pipeline - except there is no final "production" stage, and there is only one pre-production stage that executes all tests against that one environment. There is no automated rollback - developers are expected to fix-forward.
+
+6. UAT, training, and copies of the service exist on their own branches with their own copies of the pipeline subscribed to them. Changes may be merged into these branches before or after they reach master depending on the circumstances.
+
+Illustrated, we might expect this to look something like the following graphic.
+
+<center><img src="/img/manual-gates/model.png" /></center>
+<br/>
+
+This has quite a few things going for it.
+
+- We have decoupled environments that have potential manual verification (e.g. UAT) from the path to production.
+- There is a clear method for creating additional environments (e.g. UAT, dev) such that they don't impact the path to production.
+- It can support different branching models with little impact to any current delivery pipelines that may be in use.
+- Decide you don't need additional environments besides pre-prod and prod (e.g. true Trunk-Based Development)? Fine. Just don't create additional environments subscribed to branches.
+
+# Conclusion
+
+I do get it - writing tests that work well against live environment is hard. It feels like a better use of time in the early stages of a product to instead invest that time in the product itself and maybe it is. However if this is the situation you are in take the time to think tactically and make an informed decision. Review the list of possible ways to introduce automated testing, select what is likely to give you the best return-on-investment, and plan for when the best time to make that investment is. This could be a timeframe (we want to have this in place by the end of the year), when you reach a certain load (we need to consider this when we have 200 daily active users), or when your team reaches a certain size. Ideally teams should focus on reducing friction into production and that can only be achieved by removing steps that require human intervention.
+
+# References
+
```

Comment:
Link to the appropriate references from within the text - use the medium to the fullest.